### PR TITLE
Automatically generated GAGS map icons

### PR DESCRIPTION
**Pull request by ninjanomnom**
## About The Pull Request

You no longer need to create an icon for types using GAGS, instead a map icon is generated for every type that has both a config and colors specified. `icon` in such objects should refer to one of the dmi files in `icons/map_icons/` and `icon_state` needs to be the same as the typepath to refer to the correct icon in that file. Further subtypes don't need to change their icon state unless the colors or config have changed.

As all objects which use GAGS have their icon automatically replaced in init, previous values weren't being used for anything other than map icons anyway, `icon_state` however has more problems. The `icon_state` must match a state in the generated file but the file holds icons for many different objects, some of which use the same icon state. So `icon_state` must be something unique instead then, which is why the typepath is used. This isn't good for runtime code though which expects a different icon state, so I've added `post_init_icon_state` as the icon state which replaces the defined one at runtime during init.

### Note for Spriters:

After you've assigned the correct values to vars, you must run the game through init on your local machine and commit the changes to the map icon dmi files. Unit tests should catch all cases of forgetting to assign the correct vars, or not running through init.

### Note for Server Operators:

In order to not generate these icons on live I've added a new config entry which should be disabled on live called `GENERATE_ASSETS_IN_INIT` in the config.txt

Example images from sdmm:

image

image

### CODE REVIEW

**Comment by EuSouAFazer** [comment]
Heck YEAH, mapper W!

File: code/controllers/subsystem/processing/greyscale.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/processing/greyscale.dm:78]: this strikes me as highly likely to give bad icon errors due to both the check tick and the high volume of icons being processed here. can't recall if Insert requires nonstatic icons tho so maybe the problem is limited just to the CHECK_TICK

File: code/controllers/subsystem/processing/greyscale.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/processing/greyscale.dm:39]: Do you have profiles for this?

Comment (ninjanomnom) [reply code/controllers/subsystem/processing/greyscale.dm:39]: No but why does it matter, it only runs in init on dev machines. I do know it's less than 6 seconds, just from looking at subsystem init time.

File: code/controllers/subsystem/processing/greyscale.dm
Comment (ninjanomnom) [reply code/controllers/subsystem/processing/greyscale.dm:78]: I did break up the icons a bit into different files partially to avoid bad icon errors, but you're right that check tick might allow other icon operations to run and cause a bad icon error here. Want me to just remove them all since this shouldn't ever run on live anyway?

**Comment by MrMelbert** [comment]
We can reference these preview icons in other code, right? I'd love to use this for the loadout menu so we don't have a bunch of errors in the place of gags icons

**Comment by ninjanomnom** [comment]
You can use them anywhere you want just fine

File: code/controllers/subsystem/processing/greyscale.dm
Comment (LemonInTheDark) [reply code/controllers/subsystem/processing/greyscale.dm:78]: I think it'd be best to just remove checktick since it'll only ever waste time as you say, and it risks a static holder

File: code/controllers/subsystem/processing/greyscale.dm
Comment (LemonInTheDark) [reply code/controllers/subsystem/processing/greyscale.dm:39]: I care about init time. This isn't a blocker but I want to be conscious of it and profiles help me do that.

Comment (ninjanomnom) [reply code/controllers/subsystem/processing/greyscale.dm:39]: I'll grab the profiler data from the proc next time I get the chance

**Review by Wallemations — APPROVED** [review]
Damn I didn't even know you could make GAGS work with SDMM

File: code/controllers/subsystem/processing/greyscale.dm
Comment (ninjanomnom) [reply code/controllers/subsystem/processing/greyscale.dm:39]: image

Comment (LemonInTheDark) [reply code/controllers/subsystem/processing/greyscale.dm:39]: can you send me an export of these? I'd like to look at where that time goes

File: code/controllers/subsystem/processing/greyscale.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/processing/greyscale.dm:46]: am I insane or are we doing duped work here

File: code/controllers/subsystem/wardrobe.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/wardrobe.dm:256]: what?

File: code/controllers/subsystem/processing/greyscale.dm
Comment (ninjanomnom) [reply code/controllers/subsystem/processing/greyscale.dm:46]: types are skipped if they've been handled already

File: code/controllers/subsystem/wardrobe.dm
Comment (ninjanomnom) [reply code/controllers/subsystem/wardrobe.dm:256]: Previously, a lot of uninitialized clothing was being used to generate the images for stuff like antag datum icons, look at the first screenshot test comment.

Comment (ninjanomnom) [reply code/controllers/subsystem/wardrobe.dm]: I've implemented the alternate method we spoke about in the coding channel

File: code/controllers/subsystem/processing/greyscale.dm
Comment (ninjanomnom) [reply code/controllers/subsystem/processing/greyscale.dm:39]: [Profile results (total time).txt](https://github.com/user-attachments/files/16997034/Profile.results.total.time.txt)

Already sent elsewhere, but here it is for anyone else curious. Results are from clearing the profiler first before running the map icon generation.

**Comment by Wallemations** [comment]
I ain't gonna merge this since this is 100% more a code-maint thing than a sprite-maint one. I got no clue how efficient this code is.

**Comment by ninjanomnom** [comment]
This is just waiting on potato to come look it over again. I'll also need to update it one more time after it's been approved or we'll have merge skew issues.

File: code/modules/unit_tests/greyscale_map_icons.dm
Comment (ninjanomnom) [reply code/modules/unit_tests/greyscale_map_icons.dm:6]: /atom can be used by itself in theory and I don't want to leave out checking it just because it seems highly unlikely to be used.

**Review by Ghommie — APPROVED** [review]
If nobody is against some of the changes for a good reason, I'll merge it later today or tomorrow.

**Comment by ninjanomnom** [comment]
You'll need to wait for me to handle the conflicts and the many merge skew errors that popped up while waiting for review but I'll bother you or someone else for merge when that's done sometime tomorrow.

**Comment by Ghommie** [comment]
I hope you can eventually resolve the conflicts.

**Comment by timothymtorres** [comment]
I remember us talking about this years ago! Glad to see it was finally reaching maturation.

**Comment by LemonInTheDark** [comment]
sadge

**Pull request closed without merging**
